src/travertino/declaration.py

```diff
@@ -65,17 +72,168 @@ def __str__(self):
         return ", ".join(self._options)
 
 
+class validated_property:
+    def __init__(self, choices, initial=None):
+        """Define a simple validated property attribute.
+
+        :param choices: The available choices.
+        :param initial: What value to set initially. None means no initial value.
+        """
+        self.choices = choices
+        self.initial = None
+
+        try:
+            # If an initial value has been provided, it must be consistent with
+            # the choices specified.
+            if initial is not None:
+                self.initial = choices.validate(initial)
+        except ValueError:
+            # Unfortunately, __set_name__ hasn't been called yet, so we don't know the
+            # property's name.
+            raise ValueError(
+                f"Invalid initial value {initial!r}. Available choices: {choices}"
+            )
+
+    def __set_name__(self, owner, name):
+        self.name = name
+        owner._PROPERTIES[owner].add(name)
+        owner._ALL_PROPERTIES[owner].add(name)
+
+    def __get__(self, obj, objtype=None):
+        if obj is None:
+            return self
+
+        value = getattr(obj, f"_{self.name}", None)
+        return self.initial if value is None else value
+
+    def __set__(self, obj, value):
+        if value is self:
+            # This happens during autogenerated dataclass __init__ when no value is
+            # supplied.
+            return
+
+        if value is None:
+            raise ValueError(
+                "Python `None` cannot be used as a style value; "
+                f"to reset a property, use del `style.{self.name}`"
+            )
+
+        try:
+            value = self.choices.validate(value)
+        except ValueError:
+            raise ValueError(
+                f"Invalid value {value!r} for property {self.name}; "
+                f"Valid values are: {self.choices}"
+            )
+
+        if value != getattr(obj, f"_{self.name}", None):
+            setattr(obj, f"_{self.name}", value)
+            obj.apply(self.name, value)
+
+    def __delete__(self, obj):
+        try:
+            delattr(obj, f"_{self.name}")
+        except AttributeError:
+            pass
+        else:
+            obj.apply(self.name, self.initial)
+
+
+class directional_property:
+    DIRECTIONS = [TOP, RIGHT, BOTTOM, LEFT]
+    ASSIGNMENT_SCHEMES = {
```

Comment:
Cute :-)

Reply:
Might as well use what we have, right? : )